### PR TITLE
Export dynamic phenotypes with gene symbols to a separate PDF

### LINKED ISSUE #2168: Download of HPO gene panel
When downloading the generated "HPO gene panel list" you now get a long list in excel with "HGNC symbol,HGNC ID,Description" - 

When reporting our results to the referral physician we add this list as an attached document (pdf). 

To save us some time making this list - could it be possible to:

1. Have all genes in alphabetical order (now it's random/alphabetical per HPO = a mess)
2. Only the HGNC symbol
3. HPO-terms used as headline (i.e. Overgrowth HP:0001548, Macrocephaly HP:0000256)
4. Not one gene per row, but a comprehensive list of genes separated with a comma (i.e. Gene A, Gene B, Gene C...) 
5. In a pdf document

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #2168

Instead of an excel sheet, when clicking on "Download HPO gene panel":
image

- download a PDF file
- with HPO terms as headers
- with ordered genes symbols separated by comma

**How to test**:
1. Install on stage or locally ( if locally run the command `scout --demo update hpo` first)
1. Add more than one HPO terms to a case
1. Generate a dynamic gene list
1. Click on the "Download HPO gene panel"
1. Make sure that it looks OK

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by ML, DN
- [x] tests executed by CR, DN

### CODE REVIEW

**Comment by northwestwitch** [comment]
Testing with a couple of phenotype terms containing genes:

image

**Comment by northwestwitch** [comment]
Ready for review. An easy one!

**Comment by mikaell** [comment]
> Ready for review. An easy one!

I'll have a look

**Review by mikaell — COMMENTED** [review]
Looks good.

* Think you missed to comment on one argument.

File: scout/server/blueprints/cases/controllers.py [hunk 1]
```diff
@@ -586,6 +586,37 @@ def update_cancer_samples(
     )
 
 
+def phenotypes_genes(store, hpo_ids, build="37"):
+    """Generate a dictionary consisting of phenotype term with associated genes from a list of HPO IDs
+
+    Args:
+        hpo_ids(list): a list of HPO terms
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/cases/controllers.py:593]: Thanks, I'll fix it. I'm thinking to introduce also a control that genome build of the case is either "37" or "38", to force it into these two options if it's something like GRCh37 or GRCh38

**Comment by northwestwitch** [comment]
Fixed @mikaell 
I've also fixed a test to include the case when the genome build is not the standard "37"

**Comment by northwestwitch** [comment]
Hi @mikaell I've fixed the comment as you asked, could you re-review? If it's a just a comment change that you require perhaps next time you could approve directly? So it goes faster

**Comment by dnil** [comment]
Looks great - very professional! Did you check back with anyone from the clinic side (notably Anna H who wrote about it) to see that she liked it? 😊  In particular, the design choice to split into different HPO-terms looks good to me, but one may argue for and against it on account of gene symbols getting repeated and the list getting long when similar or functionally related terms are used?

**Comment by northwestwitch** [comment]
@

> Testing with a couple of phenotype terms containing genes:
> 
> image

**Comment by northwestwitch** [comment]
> Did you check back with anyone from the clinic side (notably Anna H who wrote about it) to see that she liked it?

Good idea! @AnnHam would something like this do?
image

**Comment by AnnHam** [comment]
That's really nice! Question: What happens with the gene list if I say 2 out of 4 HPO should match?


________________________________
Från: Chiara Rasi <notifications@github.com>
Skickat: den 12 november 2020 12:24
Till: Clinical-Genomics/scout <scout@noreply.github.com>
Kopia: Anna Hammarsjö <anna.hammarsjo@ki.se>; Mention <mention@noreply.github.com>
Ämne: Re: [Clinical-Genomics/scout] Export dynamic phenotypes with gene symbols to a separate PDF (#2171)


Did you check back with anyone from the clinic side (notably Anna H who wrote about it) to see that she liked it?

Good idea! @AnnHam<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAnnHam&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca8dc4dfb17e64480bd1e08d886fd7a38%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407770546735849%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nqmydDROAj3HD0IBNZh8HI8geTergPzY8QjIS5LF6vM%3D&reserved=0> would something like this do?
[image]<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F28093618%2F98934124-e26d7d80-24e1-11eb-915f-5e0f855859f8.png&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca8dc4dfb17e64480bd1e08d886fd7a38%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407770546745844%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=20mFL81I%2BAoJR9nAfcGSAcCn3CWkTgl91jc2meE7qEw%3D&reserved=0>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FClinical-Genomics%2Fscout%2Fpull%2F2171%23issuecomment-726017877&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca8dc4dfb17e64480bd1e08d886fd7a38%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407770546745844%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=lSgd4T4Nz9minkmAw7S4jq1CtspcNrfmU9R6hPPau1o%3D&reserved=0>, or unsubscribe<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHX34IHPZ2LG4NJMZV4RFTLSPPAVTANCNFSM4THHXTTQ&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca8dc4dfb17e64480bd1e08d886fd7a38%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407770546755836%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=GnkoH9OXjbFruLGC%2FO5elMO7FrthBOzYtqXDszGPUoM%3D&reserved=0>.



När du skickar e-post till Karolinska Institutet (KI) innebär detta att KI kommer att behandla dina personuppgifter. Här finns information om hur KI behandlar personuppgifter<https://ki.se/medarbetare/integritetsskyddspolicy>.


Sending email to Karolinska Institutet (KI) will result in KI processing your personal data. You can read more about KI’s processing of personal data here<https://ki.se/en/staff/data-protection-policy>.

**Comment by northwestwitch** [comment]
> That's really nice! Question: What happens with the gene list if I say 2 out of 4 HPO should match?

I'm not sure I understand the question, should match to what?

**Comment by AnnHam** [comment]
Sorry, when having more than one HPO - you can say that you want all HPO or a specific numbers of HPO to match with your gene panel - so how does a gene list of that "search" looks like? Will you still have the heading of each specific HPO, but only genes mathing with another HPO (if you have set the number to for example 2 matching HPO). I'm not sure you can follow my explanation (working from home today and I don't have Scout accessible right now to look at the interface...)



________________________________
Från: Chiara Rasi <notifications@github.com>
Skickat: den 12 november 2020 12:38
Till: Clinical-Genomics/scout <scout@noreply.github.com>
Kopia: Anna Hammarsjö <anna.hammarsjo@ki.se>; Mention <mention@noreply.github.com>
Ämne: Re: [Clinical-Genomics/scout] Export dynamic phenotypes with gene symbols to a separate PDF (#2171)


That's really nice! Question: What happens with the gene list if I say 2 out of 4 HPO should match?

I'm not sure I understand the question, should match to what?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FClinical-Genomics%2Fscout%2Fpull%2F2171%23issuecomment-726025040&data=04%7C01%7Canna.hammarsjo%40ki.se%7Cb855b3f951794868674308d886ff7de7%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407779169582736%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=t%2Fir8N7I8vkIgp8S8zXhoemd776CXskuexBNiSfNO18%3D&reserved=0>, or unsubscribe<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHX34ICMBJKNKQ6VEMCMGH3SPPCLTANCNFSM4THHXTTQ&data=04%7C01%7Canna.hammarsjo%40ki.se%7Cb855b3f951794868674308d886ff7de7%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407779169592734%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=KjIy6fO44dWBdRKBWCo0DDjlWf1AXfanNaN7yNb6qmU%3D&reserved=0>.



När du skickar e-post till Karolinska Institutet (KI) innebär detta att KI kommer att behandla dina personuppgifter. Här finns information om hur KI behandlar personuppgifter<https://ki.se/medarbetare/integritetsskyddspolicy>.


Sending email to Karolinska Institutet (KI) will result in KI processing your personal data. You can read more about KI’s processing of personal data here<https://ki.se/en/staff/data-protection-policy>.

**Comment by northwestwitch** [comment]
> so how does a gene list of that "search" looks like?

Oh yes I see now! The download is not actually taking into account the phenomizer filter (didn't think about it!). So when you have let's say those 2 hpo terms and you filter for 2 min matches you get only 9 genes in the HPO gene panel. But if you export it then it exports them all (more than 9). If this is not the intended behavior I can fix it. What should it export, only the filtered genes and empty header for the phenotypes with no genes after the filtering?

**Comment by AnnHam** [comment]
Yes, only the filtered genes. Maybe you'd like all genes together when saying 2 or more HPO should match? Is this possible?

Don't know what you mean with: "empty header for the phenotypes with no genes after the filtering"

The purpose of this is that when writing the report to the referral clinican we add the gene list of analyzed genes as an attachment and if this attachment could be a page or two that's perfect. As it is now you get a long list with one gene per row (=10 pages sometimes).
________________________________
Från: Chiara Rasi <notifications@github.com>
Skickat: den 12 november 2020 13:47
Till: Clinical-Genomics/scout <scout@noreply.github.com>
Kopia: Anna Hammarsjö <anna.hammarsjo@ki.se>; Mention <mention@noreply.github.com>
Ämne: Re: [Clinical-Genomics/scout] Export dynamic phenotypes with gene symbols to a separate PDF (#2171)


so how does a gene list of that "search" looks like?

Oh yes I see now! The download is not actually taking into account the phenomizer filter (didn't think about it!). So when you have let's say those 2 hpo terms and you filter for 2 min matches you get only 9 genes in the HPO gene panel. But if you export it then it exports them all (more than 9). If this is not the intended behavior I can fix it. What should it export, only the filtered genes and empty header for the phenotypes with no genes after the filtering?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FClinical-Genomics%2Fscout%2Fpull%2F2171%23issuecomment-726057010&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca3895921964c482799ca08d887092d33%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407820761376063%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=jhz6Zpe1s3wzUpIMyLTnHDTIcJutIbmJMTA8rCpI56Q%3D&reserved=0>, or unsubscribe<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHX34IC7VAOJRGR7Z4JNTZ3SPPKPVANCNFSM4THHXTTQ&data=04%7C01%7Canna.hammarsjo%40ki.se%7Ca3895921964c482799ca08d887092d33%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637407820761386015%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=NbZGUGNoioLWdvF38BeAIw%2FZxBLKZIjSZKQLDSBvdW0%3D&reserved=0>.



När du skickar e-post till Karolinska Institutet (KI) innebär detta att KI kommer att behandla dina personuppgifter. Här finns information om hur KI behandlar personuppgifter<https://ki.se/medarbetare/integritetsskyddspolicy>.


Sending email to Karolinska Institutet (KI) will result in KI processing your personal data. You can read more about KI’s processing of personal data here<https://ki.se/en/staff/data-protection-policy>.

**Comment by northwestwitch** [comment]
Yes, I was about to ask about it. Whenever you filter by more than one phenotype you'd get something like this, which is not really useful: 

image


>Don't know what you mean with: "empty header for the phenotypes with no genes after the filtering"
I don't know either, let's forget about it. 😆 

Ok so just a recap:
- If user doesn't filter by more than 1 phenotype show the HPO header with the genes (as it was originally)
- If the user uses the "min matches" with more than 1 phenotype then put all the phenotypes in the header and only the filtered genes in the gene list.

Right?

**Comment by northwestwitch** [comment]
Otherwise I just had this idea:
What if I leave the layout with HPO term and list of genes as it is and then if you have filtered by more than 1 phenotype I make the genes that match bold, while the rest are more light, like gray?

**Comment by northwestwitch** [comment]
Hi @AnnHam what do you think of my latest suggestion for this feature? Which way do you prefer?

**Comment by AnnHam** [comment]
Hej, we don’t want to see the genes we haven’t filtered for - only genes included in the HPO Gene list. I’m not sure we need to see which Gene belongs to which HPO (when using more than one matching HPO).

Hämta Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Chiara Rasi <notifications@github.com>
Skickat: Wednesday, November 18, 2020 8:25:18 AM
Till: Clinical-Genomics/scout <scout@noreply.github.com>
Kopia: Anna Hammarsjö <anna.hammarsjo@ki.se>; Mention <mention@noreply.github.com>
Ämne: Re: [Clinical-Genomics/scout] Export dynamic phenotypes with gene symbols to a separate PDF (#2171)


Hi @AnnHam<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAnnHam&data=04%7C01%7Canna.hammarsjo%40ki.se%7C376aef9e71944990d76408d88b931aa5%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637412811204102473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=81w0SsaijpavXxmo9pw7%2BnO%2B66VXL5XgZGhTRlQilN4%3D&reserved=0> what do you think of my latest suggestion for this feature? Which way do you prefer?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FClinical-Genomics%2Fscout%2Fpull%2F2171%23issuecomment-729490399&data=04%7C01%7Canna.hammarsjo%40ki.se%7C376aef9e71944990d76408d88b931aa5%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637412811204112467%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=JeGZQXNaEhJsWJ2w7DU1nvgJtgVMtfJOBZR7m1%2FqxJY%3D&reserved=0>, or unsubscribe<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHX34IDFLXMXT2NV43X3TZ3SQNZF5ANCNFSM4THHXTTQ&data=04%7C01%7Canna.hammarsjo%40ki.se%7C376aef9e71944990d76408d88b931aa5%7Cbff7eef1cf4b4f32be3da1dda043c05d%7C0%7C0%7C637412811204112467%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=vf%2FjTdMMA7Z9o0%2F8E9BeojE62jXZ5Q%2BhnJEFigysA8U%3D&reserved=0>.



När du skickar e-post till Karolinska Institutet (KI) innebär detta att KI kommer att behandla dina personuppgifter. Här finns information om hur KI behandlar personuppgifter<https://ki.se/medarbetare/integritetsskyddspolicy>.


Sending email to Karolinska Institutet (KI) will result in KI processing your personal data. You can read more about KI’s processing of personal data here<https://ki.se/en/staff/data-protection-policy>.

**Comment by northwestwitch** [comment]
> Hej, we don’t want to see the genes we haven’t filtered for - only genes included in the HPO Gene list. I’m not sure we need to see which Gene belongs to which HPO (when using more than one matching HPO). Hämta Outlook för iOS<https://aka.ms/o0ukef>

Makes sense, I will fix it that way!

**Comment by northwestwitch** [comment]
Modified according to the latest feedback. I've installed the branch on Scout stage and tested the 2 scenarios:

- ### Multiple assigned phenotype without filtering for genes matching 2 or more phenotypes:

Example: 
image

Results of the PDF export:

image

- ### Multiple assigned phenotypes, creating HPO panel by filtering by 2 phenotypes:
image

Results of the PDF export:
image


I think this is the intended behavior now. 

Could someone re-review the code so we can merge it and have the feature in the new release?

**Comment by northwestwitch** [comment]
@mikaell or @dnil whenever you have some time I could use a new review for this one. Thanks!

**Comment by mikaell** [comment]
> @mikaell or @dnil whenever you have some time I could use a new review for this one. Thanks!

I can do this during the day.

File: scout/server/blueprints/cases/controllers.py [hunk 2]
```diff
@@ -586,6 +586,61 @@ def update_cancer_samples(
...
+            continue
+        # Create a list with all gene symbols (or HGNC ID if symbol is missing) associated with the phenotype
+        gene_list = []
+        for gene_id in hpo_term.get("genes", []):
+            gene_obj = store.hgnc_gene(gene_id, build)
+            if gene_obj is None:
+                continue
+            if gene_id not in dynamic_gene_list:
+                # gene was filtered out because min matching phenotypes > 1
+                by_phenotype = False  # do not display genes by phenotype
```
Comment (dnil) [inline scout/server/blueprints/cases/controllers.py:624]: 👍

**Review by dnil — APPROVED** [review]
Will be much appreciated! I have a strong feeling we will want to do the same for the gene panels, but one thing at the time.

File: tests/server/blueprints/cases/test_cases_views.py [hunk 3]
```diff
@@ -434,7 +434,7 @@ def test_download_hpo_genes(app, case_obj, institute_obj):
         # THEN the response should be successful
         assert resp.status_code == 200
         # And should download a txt file
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_views.py:436]: Right! :)

File: tests/server/blueprints/cases/test_cases_controllers.py [hunk 4]
```diff
@@ -2,7 +2,74 @@
...
+    phenotypes_genes,
+)
+
+
+def test_phenotypes_genes(gene_database, case_obj):
+    """Test function that creates phenotype terms dictionaries with gene symbol info"""
+
+    adapter = gene_database
+    # Given a database with one phenotype term containing genes
+    gene_list = [26113, 9479, 10889, 18040, 10258]
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_controllers.py:18]: The list is used by this file only. Would it be OK to create a constant on top of the file:

GENE_LIST = [26113, 9479, 10889, 18040, 10258]

And use that instead?

File: scout/server/blueprints/cases/views.py [hunk 5]
```diff
@@ -915,30 +915,17 @@ def download_hpo_genes(institute_id, case_name):
...
-
-    headers.add(
-        "Content-Disposition",
-        "attachment",
-        filename=f"HPO_gene_list.{institute_id}.{case_name}.{download_day}.csv",
-    )
-    return Response(
-        _generate_csv(",".join(csv_header), csv_lines), mimetype="text/csv", headers=headers
+    # Create export object consisting of dynamic phenotypes with associated genes as a dictionary
+    phenotype_genes = controllers.phenotypes_genes(store, case_obj)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/cases/views.py:919]: Yes. much better!

**Comment by northwestwitch** [comment]
> I can do this during the day.

Done, but thanks anyway! 🙏

File: tests/server/blueprints/cases/test_cases_controllers.py [hunk 6]
```diff
@@ -2,7 +2,74 @@
...
+
+
+def test_phenotype_genes_matching_phenotypes(gene_database, case_obj):
+    """Test the function that creates the phenotype-genes terms dictionaries with genes matching more than 1 phenotype"""
+
+    adapter = gene_database
+    assert adapter.hgnc_collection.find_one()
+    # Given a database with 2 phenotype term containing matching genes
+    gene_list = [26113, 9479, 10889, 18040, 10258]
+
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_controllers.py:52]: Yeah I've put it in the constant on top. We could move it to the fixtures if we need something like that in some other test on another file!

File: tests/server/blueprints/cases/test_cases_controllers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dnil) [reply tests/server/blueprints/cases/test_cases_controllers.py]: Sure, just avoid the repetition and allow easy reuse of especially the reused hpo term, and the other overlapping one - but why not make them fixtures at least locally already? You could keep them in the same file, or most detailed level conftest  for this one.

[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_controllers.py]: Done! I've moved both hpo_term and gene_list to the local fixtures now! Thanks for the review!

**Pull request merged**